### PR TITLE
[BLE] Change BLE to use zjs callback API

### PR DESCRIPTION
**Pull request by jimmy-huang**
Signed-off-by: Jimmy Huang <jimmy.huang@intel.com>

### CODE REVIEW

**Comment by jimmy-huang** [comment]
@jprestwo thanks, updated PR

**Comment by jprestwo** [comment]
+1 Looks good

File: src/zjs_ble.c [hunk 1]
```diff
@@ -285,52 +283,45 @@ static jerry_value_t zjs_ble_write_attr_call_function_return(const jerry_value_t
...
-           } else {
-               args[0] = jerry_create_null();
+               buf->bufsize == cb->buffer_size) {
+               memcpy(buf->buffer, cb->buffer, cb->buffer_size);
            }
-        } else {
-            args[0] = jerry_create_null();
+
+           jerry_release_value(args[0]);
+           args[0]= buf_obj;
```
Comment (grgustaf) [inline src/zjs_ble.c]: space before equals

File: src/zjs_ble.c [hunk 2]
```diff
@@ -285,52 +283,45 @@ static jerry_value_t zjs_ble_write_attr_call_function_return(const jerry_value_t
...
-    rval = jerry_call_function(mycb->zjs_cb.js_callback, chrc->chrc_obj, args, 4);
+    rval = jerry_call_function(cb->js_callback, chrc->chrc_obj, args, 4);
     if (jerry_value_has_error_flag(rval)) {
-        PRINT("zjs_ble_write_attr_call_function: failed to call onWriteRequest function\n");
+        PRINT("zjs_ble_write_c_callback: failed to call onWriteRequest function\n");
     }
 
     jerry_release_value(args[0]);
-    jerry_release_value(args[1]);
-    jerry_release_value(args[2]);
```
Comment (grgustaf) [inline src/zjs_ble.c]: Why did you remove these two?

File: src/zjs_ble.c [hunk 3]
```diff
@@ -405,28 +395,55 @@ static jerry_value_t zjs_ble_update_value_call_function(const jerry_value_t func
...
-    args[1] = jerry_create_external_function(zjs_ble_update_value_call_function);
-    rval = jerry_call_function(mycb->zjs_cb.js_callback, chrc->chrc_obj, args, 2);
+    args[1] = jerry_create_external_function(zjs_ble_update_value_callback_function);
+    rval = jerry_call_function(cb->js_callback, chrc->chrc_obj, args, 2);
     if (jerry_value_has_error_flag(rval)) {
-        PRINT("zjs_ble_subscribe_call_function: failed to call onSubscribe function\n");
+        PRINT("zjs_ble_subscribe_c_callback: failed to call onSubscribe function\n");
     }
 
-    jerry_release_value(args[0]);
```
Comment (grgustaf) [inline src/zjs_ble.c]: Why?

File: src/zjs_ble.c [hunk 4]
```diff
@@ -405,28 +395,55 @@ static jerry_value_t zjs_ble_update_value_call_function(const jerry_value_t func
...
+
+    jerry_value_t rval;
+
+    rval = jerry_call_function(cb->js_callback, chrc->chrc_obj, NULL, 0);
+    if (jerry_value_has_error_flag(rval)) {
+        PRINT("zjs_ble_notify_c_callback: failed to call onNotify function\n");
+    }
+
+    jerry_release_value(rval);
+}
```
Comment (grgustaf) [inline src/zjs_ble.c]: These two functions are identical, I think, so make a more generic name and use it twice?

File: src/zjs_ble.c [hunk 5]
```diff
@@ -855,6 +887,11 @@ static bool zjs_ble_parse_service(ble_service_t *service)
         }
 
         memset(chrc, 0, sizeof(ble_characteristic_t));
+        chrc->read_cb.id = chrc->write_cb.id
+                         = chrc->subscribe_cb.id
+                         = chrc->unsubscribe_cb.id
+                         = chrc->notify_cb.id
+                         = -1;
```
Comment (grgustaf) [inline src/zjs_ble.c]: Hmm, seems wacky. But if that's what the kids are doing these days. :)

**Comment by grgustaf** [comment]
Also, two unrelated patches from @haoxli are currently in the PR?

File: src/zjs_ble.c (hunk 2 above)
Comment (jimmy-huang) [inline src/zjs_ble.c]: argp[1] and arg[2] are integer and boolean, so they do not need to be released, but I'll put them back in for the sake in the case the jerryscript api changes in the future

File: src/zjs_ble.c (hunk 5 above)
Comment (jimmy-huang) [inline src/zjs_ble.c]: if there's a better/cleaner way to do this, let me know, i can adapt.

File: src/zjs_ble.c (hunk 4 above)
Comment (jimmy-huang) [inline src/zjs_ble.c]: they are not exactly the same, since the first *cb points to &chrc->unsubscribe_cb, but the second one points to &chrc->notify_cb.
